src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+        {
+            AsyncCompletionData.UpdateSelectionHint selection;
+            if (triggerReason == AsyncCompletionData.CompletionTriggerReason.Insertion)
+            {
+                // If the user was just typing, and the list went to empty *and* this is a 
+                // language that wants to dismiss on empty, then just return a null model
+                // to stop the completion session.
+                if (completionRules.DismissIfEmpty)
+                {
+                    return null;
+                }
+
+                selection = AsyncCompletionData.UpdateSelectionHint.NoChange;
+            }
+            else
+            {
+                // If the user has turned on some filtering states, and we filtered down to
+                // nothing, then we do want the UI to show that to them.  That way the user
+                // can turn off filters they don't want and get the right set of items.
+
+                // If we are going to filter everything out, then just preserve the existing
+                // model (and all the previously filtered items), but switch over to soft
+                // selection.
+                selection = hasActiveFilters ? AsyncCompletionData.UpdateSelectionHint.NoChange : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight>.Empty, selectedItemIndex: 0,
+                filters, selection, centerSelection: true, uniqueItem: default);
+        }
+
+        private IEnumerable<AsyncCompletionData.CompletionItemWithHighlight> GetHighlightedList(
+            IEnumerable<ExtendedFilterResult> filterResults,
+            string filterText,
+            bool highlightMatchingPortions)
+        {
+            var highlightedList = new List<AsyncCompletionData.CompletionItemWithHighlight>();
+            foreach (var item in filterResults)
+            {
+                var highlightedSpans = highlightMatchingPortions
+                                        ? _completionHelper.GetHighlightedSpans(item.VSCompletionItem.DisplayText, filterText, CultureInfo.CurrentCulture)
+                                        : ImmutableArray<TextSpan>.Empty;
+                highlightedList.Add(new AsyncCompletionData.CompletionItemWithHighlight(item.VSCompletionItem, highlightedSpans.Select(s => s.ToSpan()).ToImmutableArray()));
+            }
+
+            return highlightedList;
+        }
+
+        private static ImmutableArray<AsyncCompletionData.CompletionFilterWithState> GetUpdatedFilters(
+            List<ExtendedFilterResult> filteredList,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters)
+        {
+            // See which filters might be enabled based on the typed code
+            var textFilteredFilters = filteredList.SelectMany(n => n.VSCompletionItem.Filters).Distinct();
```

Comment:
i believe .Distinct is lazy.  So ecah time you call .Contains below you're computing this entire list *every time*.